core/chains/evm/client/tx_simulator.go

```diff
@@ -0,0 +1,134 @@
+	}
+	// ClassifySendError will not have the proper fields for logging within the method due to the empty Transaction passed
+	code := ClassifySendError(err, lggr, &types.Transaction{}, msg.From, chainType.IsL2())
+	// Only return error if ZK OOC error is identified
+	if code == commonclient.OutOfCounters {
+		return errors.New(ErrOutOfCounters)
+	}
+	return nil
+}
+
+// eth_estimateGas returns out-of-counters (OOC) error if the transaction would result in an overflow
+func simulateTransactionDefault(ctx context.Context, client simulatorClient, msg ethereum.CallMsg) error {
+	var result hexutil.Big
+	errCall := client.CallContext(ctx, &result, "eth_estimateGas", toCallArg(msg), "pending")
+	jsonErr, _ := ExtractRPCError(errCall)
+	if jsonErr != nil && len(jsonErr.Message) > 0 {
+		return errors.New(jsonErr.Message)
+	}
+	return nil
+}
+
+type zkEvmEstimateCountResponse struct {
+	CountersUsed struct {
+		GasUsed              string
+		UsedKeccakHashes     string
+		UsedPoseidonHashes   string
+		UsedPoseidonPaddings string
+		UsedMemAligns        string
+		UsedArithmetics      string
+		UsedBinaries         string
+		UsedSteps            string
+		UsedSHA256Hashes     string
+	}
+	CountersLimit struct {
+		MaxGasUsed          string
+		MaxKeccakHashes     string
+		MaxPoseidonHashes   string
+		MaxPoseidonPaddings string
+		MaxMemAligns        string
+		MaxArithmetics      string
+		MaxBinaries         string
+		MaxSteps            string
+		MaxSHA256Hashes     string
+	}
+	OocError string
+}
+
+// zkEVM implemented a custom zkevm_estimateCounters method to detect if a transaction would result in an out-of-counters (OOC) error
+func simulateTransactionZkEvm(ctx context.Context, client simulatorClient, lggr logger.SugaredLogger, msg ethereum.CallMsg) error {
+	var result zkEvmEstimateCountResponse
+	err := client.CallContext(ctx, &result, "zkevm_estimateCounters", toCallArg(msg), "pending")
+	if err != nil {
+		return fmt.Errorf("failed to simulate tx: %w", err)
+	}
+	if detectZkEvmCounterOverflow(result) && len(result.OocError) > 0 {
+		lggr.Debugw("zkevm_estimateCounters returned", "result", result)
+		return errors.New(result.OocError)
+	}
+	return nil
+}
+
+// Helper method for zkEvm to determine if response indicates an overflow
+func detectZkEvmCounterOverflow(result zkEvmEstimateCountResponse) bool {
+	if result.CountersUsed.UsedKeccakHashes > result.CountersLimit.MaxKeccakHashes ||
+		result.CountersUsed.UsedPoseidonHashes > result.CountersLimit.MaxPoseidonHashes ||
+		result.CountersUsed.UsedPoseidonPaddings > result.CountersLimit.MaxPoseidonPaddings ||
+		result.CountersUsed.UsedMemAligns > result.CountersLimit.MaxMemAligns ||
+		result.CountersUsed.UsedArithmetics > result.CountersLimit.MaxArithmetics ||
+		result.CountersUsed.UsedBinaries > result.CountersLimit.MaxBinaries ||
+		result.CountersUsed.UsedSteps > result.CountersLimit.MaxSteps ||
+		result.CountersUsed.UsedSHA256Hashes > result.CountersLimit.MaxSHA256Hashes {
```

Comment:
Comparing these hex values lexicographically will not work as `"0x9">"0x123" == true`. Setting the fields to e.g. `hexutil.Uint64` should unmarshal them correctly.
Tests should have caught that, as your input contains that case but the `&&` in `simulateTransactionZkEvm ` masked it.

Reply:
Shoot good catch! I changed around the response types a few times during dev and forgot these aren't already uint64 😅. But I'll wait to address this once we decide whether we want this method or not in the other comment.